lib/AST/DiagnosticConsumer.cpp

```diff
@@ -163,41 +157,78 @@ void FileSpecificDiagnosticConsumer::handleDiagnostic(
     StringRef FormatString, ArrayRef<DiagnosticArgument> FormatArgs,
     const DiagnosticInfo &Info) {
 
-  Optional<DiagnosticConsumer *> specificConsumer;
+  Optional<ConsumerSpecificInformation *> consumerSpecificInfo;
   switch (Kind) {
   case DiagnosticKind::Error:
   case DiagnosticKind::Warning:
   case DiagnosticKind::Remark:
-    specificConsumer = consumerForLocation(SM, Loc);
-    ConsumerForSubsequentNotes = specificConsumer;
+    consumerSpecificInfo = consumerSpecificInformationForLocation(SM, Loc);
+    ConsumerSpecificInfoForSubsequentNotes = consumerSpecificInfo;
     break;
   case DiagnosticKind::Note:
-    specificConsumer = ConsumerForSubsequentNotes;
+    consumerSpecificInfo = ConsumerSpecificInfoForSubsequentNotes;
     break;
   }
-
-  if (!specificConsumer.hasValue()) {
+  if (!consumerSpecificInfo.hasValue()) {
     for (auto &subConsumer : SubConsumers) {
       if (subConsumer.second) {
         subConsumer.second->handleDiagnostic(SM, Loc, Kind, FormatString,
                                              FormatArgs, Info);
       }
     }
-  } else if (DiagnosticConsumer *c = specificConsumer.getValue())
-    c->handleDiagnostic(SM, Loc, Kind, FormatString, FormatArgs, Info);
-  else
-    ; // Suppress non-primary diagnostic in batch mode.
+    return;
+  }
+  if (!consumerSpecificInfo.getValue()->consumer) {
+    WasAnErrorSuppressed =
+        true; // Suppress non-primary diagnostic in batch mode.
+    return;
+  }
+  consumerSpecificInfo.getValue()->consumer->handleDiagnostic(
+      SM, Loc, Kind, FormatString, FormatArgs, Info);
+  consumerSpecificInfo.getValue()->hasAnErrorBeenEmitted |=
+      Kind == DiagnosticKind::Error;
 }
 
-bool FileSpecificDiagnosticConsumer::finishProcessing() {
+bool FileSpecificDiagnosticConsumer::finishProcessing(SourceManager &SM) {
+  addNonSpecificErrors(SM);
+
   // Deliberately don't use std::any_of here because we don't want early-exit
   // behavior.
+
   bool hadError = false;
   for (auto &subConsumer : SubConsumers)
-    hadError |= subConsumer.second && subConsumer.second->finishProcessing();
+    hadError |= subConsumer.second && subConsumer.second->finishProcessing(SM);
   return hadError;
 }
 
+static void produceNonSpecificError(DiagnosticConsumer *consumer,
+                                    SourceManager &SM) {
+  Diagnostic diagnostic(
+      diag::error_some_error_occured_in_a_file_that_was_used_by_this_one);
+
+  // Stolen from DiagnosticEngine::emitDiagnostic
+  DiagnosticInfo Info;
+  Info.ID = diagnostic.getID();
+  Info.Ranges = diagnostic.getRanges();
+  Info.FixIts = diagnostic.getFixIts();
```

Comment:
We know there are no ranges or fix-its here, so it's probably better just to omit this.